README
Osadmin version 0.01
====================

Performance testing - perl vs. c++

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   ./benchmark.pl
   make install

if you want it installed to a local folder than do:

    perl Makefile.PL INSTALL_BASE=/home/user/your/local/folder
    make
    make test
    ./benchmark.pl
    make install

DEPENDENCIES

c++ compiler, g++ works fine

COPYRIGHT AND LICENCE

Copyright (C) 2014 by Magyarevity Miroszlav

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.

Except for the typemap code perlobject.map, which is copyright 1996 Dean Roehrich

I'm not sure about the license terms on perlobject.map, you'll have to see about
that yourself



Changes
Revision history for Perl extension Osadmin.

0.01  Thu May 15 16:13:19 2014
	- original version; created by h2xs 1.23 with options
		-A -n Osadmin
    - benchmark.pl shows speed and memory differences



benchmark.pl
#!/usr/bin/perl -w
use strict;
use warnings;

use blib;

use Benchmark ':hireswallclock';
use Osadmin_XS;
use Osadmin_PP;
use Osadmin::Benchmark;

use Data::Dumper;

my $dict_fp = '/usr/share/dict/words';

my $xs = Osadmin_XS->new($dict_fp);
my $pp = Osadmin_PP->new($dict_fp);

my $ob = Osadmin::Benchmark->new();
$ob->mem_usage_print('memory usage start');


my $do_other = 1;

# on my box, perl takes 35 MB RSS to load words, c++ requires 10MB RSS
 
$ob->mds_clean;
$pp->dict_load_words();
$ob->mem_dif_print('mem dif after pp load');

$ob->mds_clean;
$xs->dict_load_words();
$ob->mem_dif_print('mem dif after xs load');

# lets time them
# my box says c++ is twice as fast as perl loading data from file

$ob->mds_clean;
timethese (
    -10, 
    {
        'c++ load words' => sub { my $xs = Osadmin_XS->new($dict_fp); $xs->dict_load_words(); },
        'pp  load words' => sub { my $pp = Osadmin_PP->new($dict_fp); $pp->dict_load_words(); },
    }
)
    if $do_other;

$ob->mem_dif_print('mem dif after 10 second load words test');

# on my box, c++ is 14 times faster than perl at vowel calc

# the difference is huge on my box, c++ has 139.3 iterations/sec while perl has 2.72

$ob->mds_clean;
timethese (
    -10,
    {
        'c++ palindrome' => sub { $xs->palindrome_calc(); },
        'pp  palindrome' => sub { $pp->palindrome_calc(); },
    }
);

$ob->mem_dif_print('mem dif after 10 second palindrome test');

$ob->mds_clean;
timethese (
    -10, 
    {
        'pp  vowels' => sub { $pp->vowel_calc(); },
        'c++ vowels' => sub { $xs->vowel_calc(); },
    }
);
$ob->mem_dif_print('mem dif after 10 second vowels test');

print "INFO: dumping perl and c++ structure for the word level\n";

print Dumper $pp->word_get_hr('level');
print Dumper $xs->word_get_hr('level');

# c++ is almost 2 times slower since it has to create perl hash ref each time
# perl just returns a reference ;)

$ob->mds_clean;
timethese (
    -10,
    {
        'c++ word hash ref' => sub { $xs->word_get_hr('marathon'); },
        'pp  word hash ref' => sub { $pp->word_get_hr('marathon'); },
    }
)  if $do_other;
$ob->mem_dif_print('mem dif after 10 second word get hr, c++ had to create perl hashref each time');



exit;

__END__

my results:

memory usage start
       vsz	       rss	       sha	       txt	       lib	       dat	       drt
     59968	      7260	      2612	         8	         0	      4860	         0
mem dif after pp load
       vsz	       rss	       sha	       txt	       lib	       dat	       drt
     35212	     35092	        40	         0	         0	     35212	         0
mem dif after xs load
       vsz	       rss	       sha	       txt	       lib	       dat	       drt
     10824	     10824	        60	         0	         0	     10824	         0
Benchmark: running c++ load words, pp  load words for at least 10 CPU seconds...
c++ load words: 10.9024 wallclock secs ( 9.98 usr +  0.13 sys = 10.11 CPU) @  5.54/s (n=56)
pp  load words: 10.6402 wallclock secs (10.26 usr +  0.10 sys = 10.36 CPU) @  2.12/s (n=22)
mem dif after 10 second load words test
       vsz	       rss	       sha	       txt	       lib	       dat	       drt
     28976	     28404	        44	         0	         0	     28976	         0
Benchmark: running c++ palindrome, pp  palindrome for at least 10 CPU seconds...
c++ palindrome: 10.6553 wallclock secs (10.47 usr +  0.08 sys = 10.55 CPU) @ 141.33/s (n=1491)
pp  palindrome: 10.2845 wallclock secs (10.13 usr +  0.06 sys = 10.19 CPU) @  2.45/s (n=25)
mem dif after 10 second palindrome test
       vsz	       rss	       sha	       txt	       lib	       dat	       drt
      2308	      2904	         0	         0	         0	      2308	         0
Benchmark: running c++ vowels, pp  vowels for at least 10 CPU seconds...
c++ vowels: 10.6253 wallclock secs (10.45 usr +  0.08 sys = 10.53 CPU) @ 67.62/s (n=712)
pp  vowels: 10.4902 wallclock secs (10.33 usr +  0.07 sys = 10.40 CPU) @  4.71/s (n=49)
mem dif after 10 second vowels test
       vsz	       rss	       sha	       txt	       lib	       dat	       drt
         0	         0	         0	         0	         0	         0	         0
INFO: dumping perl and c++ structure for the word level
$VAR1 = {
          'vowels' => 2,
          'palindrome' => 1
        };
$VAR1 = {
          'palindrome' => 1,
          'vowels' => 2
        };
Benchmark: running c++ word hash ref, pp  word hash ref for at least 10 CPU seconds...
c++ word hash ref: 10.5404 wallclock secs (10.45 usr +  0.03 sys = 10.48 CPU) @ 469006.68/s (n=4915190)
pp  word hash ref: 10.8767 wallclock secs (10.84 usr +  0.04 sys = 10.88 CPU) @ 683223.81/s (n=7433475)
mem dif after 10 second word get hr, c++ had to create perl hashref each time
       vsz	       rss	       sha	       txt	       lib	       dat	       drt
         0	         0	         0	         0	         0	         0	         0



lib/Osadmin_PP.pm
package Osadmin_PP;

use 5.010001;
#use 5.018002;
use strict;
use warnings;
use IO::File;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Osadmin_PP ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';

# Preloaded methods go here.

sub new {
    my ($this, $dict_fp) = @_;

    my $class = ref($this) || $this;
    my $self = {
        'dict_fp' => $dict_fp,
    };
    bless $self, $class;

    $self->{'w'} = {};

    return $self;
}

sub dict_load_words {
    my ($self) = @_;

    my $fh = IO::File->new($self->{'dict_fp'}, 'r');

    if (defined $fh) {
        
        my $w = $self->{'w'};

        while (<$fh>) {
            chomp;
            $w->{$_} = {
                'vowels'  => 0,
                'palindrome' => 0,
            };
        }
        $fh->close;

    } else {
        die "ERROR: problem loading $self->{'dict_fp'}\n";
    }
    return 1;    
}

# transliteration should be fast

sub vowel_calc {
    my ($self) = @_;

    my $w = $self->{'w'};

    foreach my $word (keys %$w) {
        $w->{$word}->{'vowels'} = $word =~ tr/aeiouAEIOU//;
    }
    return 1;
}

sub word_get_hr {
    my ($self, $w) = @_;

    return exists($self->{'w'}->{$w})?$self->{'w'}->{$w}:{};
}

sub palindrome_calc {
    my ($self) = @_;

    my $w = $self->{'w'};

    foreach my $word (keys %$w) {
        my $f    = 1;
        my $wl   = length $word;
        my $wlmo = $wl - 1 ;
        for (my $i=0; $i<int($wl/2); $i++) {
            if (substr($word, $i,1) ne substr($word, $wlmo - $i, 1)) {
                $f = 0;
                last;
            }
        }
        $w->{$word}->{'palindrome'} = $f;
    }
    return 1;
}

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Osadmin_PP - pure perl module

=head1 SYNOPSIS

  use Osadmin_PP;
  blah blah blah

=head1 DESCRIPTION

Performance testing - perl vs. c++

requires a c++ compiler, g++ works fine

=head2 EXPORT

None by default.

=head1 AUTHOR

Magyarevity Miroszlav, E<lt>perl@osadmin.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014 by Magyarevity Miroszlav

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.

Except for the typemap code perlobject.map, which is copyright 1996 Dean Roehrich

I'm not sure about the license terms on perlobject.map, you'll have to see about
that yourself

=cut


lib/Osadmin_XS.pm
package Osadmin_XS;

use 5.010001;
#use 5.018002;
use strict;
use warnings;
use Osadmin;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Osadmin_XS ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';

#require XSLoader;
#XSLoader::load('Osadmin', $VERSION);

# Preloaded methods go here.

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Osadmin_XS - uses C++ code via xs

=head1 SYNOPSIS

  use Osadmin_XS;
  blah blah blah

=head1 DESCRIPTION

Performance testing - perl vs. c++

requires a c++ compiler, g++ works fine

=head2 EXPORT

None by default.

=head1 AUTHOR

Magyarevity Miroszlav, E<lt>perl@osadmin.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014 by Magyarevity Miroszlav

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.

Except for the typemap code perlobject.map, which is copyright 1996 Dean Roehrich

I'm not sure about the license terms on perlobject.map, you'll have to see about
that yourself

=cut


lib/Osadmin/Benchmark.pm
package Osadmin::Benchmark;

use 5.010001;
#use 5.018002;
use strict;
use warnings;
use IO::File;
use POSIX;
use Data::Dumper;

require Exporter;

our @ISA = qw(Exporter);

our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';

sub new {
    my ($this) = @_;

    my $class = ref($this) || $this;
    my $self = {
        'mem_fld'   => [qw| vsz rss sha txt lib dat drt |],
        'form_n'    => "%10d\t%10d\t%10d\t%10d\t%10d\t%10d\t%10d\n",
        'page_size' => POSIX::sysconf(POSIX::_SC_PAGESIZE),
    };
    bless $self, $class;

    $self->{'form_s'} = $self->{'form_n'};
    $self->{'form_s'} =~ tr/d/s/;

    $self->mem_clean();

    return $self;
}

sub mem_clean {
    my ($self) = @_;

    $self->{'mds'} = undef;
    $self->{'mdf'} = undef;
    $self->{'mem_dif'} = undef;

    return 1;
}

sub mem_usage {
    my ($self) = @_;

    my $fp = "/proc/$$/statm";

    my $fh = IO::File->new($fp, 'r');

    return
        unless defined $fh;

    my $l = <$fh>;
    chomp $l;
    $fh->close;

    my @s = map { int $_} split ' ', $l;
    my $tr = {};
    for (my $i=0; $i<@s; $i++) {
        $tr->{ $self->{'mem_fld'}->[$i] } = $s[$i];
    }
    return $tr;
}

sub mds_clean {
    my ($self) = @_;
    $self->mem_clean;
    $self->mds;
    return 1;
}

sub mds {
    my ($self) = @_;
    $self->{'mds'} = $self->mem_usage();
    return 1;
}

sub mdf {
    my ($self) = @_;
    $self->{'mdf'} = $self->mem_usage();
    return 1;
}

sub mem_dif {
    my ($self) = @_;

    return 0
        unless defined $self->{'mds'};

    $self->mdf()
        unless defined $self->{'mdf'};

    my $mds = $self->{'mds'};
    my $mdf = $self->{'mdf'};

    my $tr = {};

    foreach my $col (@{$self->{'mem_fld'}}) {
        $tr->{$col} = $mdf->{$col} - $mds->{$col};
    }
    $self->{'mem_dif'} = $tr;
    return 1;
}

sub mem_dif_print {
    my ($self, $title) = @_;

    return 0
        unless $self->mem_dif();

    $self->mem_usage_print( $title, $self->{'mem_dif'} );
    return 1;
}

sub mem_usage_print {
    my ($self, $title, $mu) = @_;

    print "$title\n"
        if defined $title;

    $mu = $self->mem_usage()
        unless defined $mu;


    printf $self->{'form_s'}, @{$self->{'mem_fld'}};
    printf $self->{'form_n'}, map {int ($self->{'page_size'} * $_ / 1024) } @$mu{ @{$self->{'mem_fld'}} };
    return 1;
}

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Osadmin::Benchmark - Custom Benchmark module

=head1 SYNOPSIS

  use Osadmin::Benchmark;
  blah blah blah

=head1 DESCRIPTION

Custom Benchmark module

=head2 EXPORT

None by default.

=head1 AUTHOR

Magyarevity Miroszlav, E<lt>perl@osadmin.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014 by Magyarevity Miroszlav

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.

Except for the typemap code perlobject.map, which is copyright 1996 Dean Roehrich

I'm not sure about the license terms on perlobject.map, you'll have to see about
that yourself

=cut


lib/Osadmin.pm
package Osadmin;

use 5.010001;
#use 5.018002;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Osadmin ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';

require XSLoader;
XSLoader::load('Osadmin', $VERSION);

# Preloaded methods go here.

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Osadmin - A simple example on how to call C++ code from perl

=head1 SYNOPSIS

  use Osadmin;
  blah blah blah

=head1 DESCRIPTION

A simple example on how to call C++ code from perl

requires a c++ compiler, g++ works fine

=head2 EXPORT

None by default.

=head1 AUTHOR

Magyarevity Miroszlav, E<lt>perl@osadmin.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014 by Magyarevity Miroszlav

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.

Except for the typemap code perlobject.map, which is copyright 1996 Dean Roehrich

I'm not sure about the license terms on perlobject.map, you'll have to see about
that yourself

=cut
